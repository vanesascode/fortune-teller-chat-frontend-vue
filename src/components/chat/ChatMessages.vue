<template>
  <div class="flex-1 overflow-y-auto p-4">
    <div class="flex flex-col space-y-2">
      <ChatBubble v-for="message in messages" :key="message.id" v-bind="message" />
      <!-- :message="message.message"
        :isMine="message.isMine"
        :image="message.image" -->
    </div>
  </div>
</template>

<script setup lang="ts">
import type { ChatMessage } from '@/interfaces/chat-message.interface'
import ChatBubble from './ChatBubble.vue'

interface Props {
  messages: ChatMessage[]
}

defineProps<Props>()
</script>
