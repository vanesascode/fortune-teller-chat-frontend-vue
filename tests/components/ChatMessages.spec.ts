import type { ChatMessage } from '@/interfaces/chat-message.interface';
import { test, expect, describe, vi } from 'vitest';
import { mount } from '@vue/test-utils';
import ChatMessages from '@/components/chat/ChatMessages.vue';
import ChatBubble from '@/components/chat/ChatBubble.vue';

const messages: ChatMessage[] = [
  {
    id: 1,
    message: 'Hello, World!',
    isMine: true,
  },
  {
    id: 2,
    message: 'Bye, World!',
    isMine: false,
  },
  {
    id: 3,
    message: 'Hello, World!',
    isMine: true,
  },
  {
    id: 4,
    message: 'Bye, World!',
    isMine: false,
  },
];

const wrapper = mount(ChatMessages, {
  props: {
    messages,
  },
});

describe('<ChatMessages tests/>', () => {
  test('ChatMessages renders correctly', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test('Find elements by classes and ids', () => {
    expect(wrapper.find('#messagesUp').exists()).toBeTruthy();
    expect(wrapper.find('.bg-blue-200').exists()).toBeTruthy();
  });

  test('renders chat messages correctly', () => {
    const chatBubbles = wrapper.findAllComponents(ChatBubble);
    expect(chatBubbles).toHaveLength(messages.length);
  });

  test('scrolls down to the bottom when new messages are added', async () => {
    //(Alternative to solution in ChatView.spec.ts, to mock the scrolling behavior during testing.):This is to avoid the TypeError: chatRef.value.scrollTo is not a function:
    const scrollToMock = vi.fn();
    console.log(wrapper.vm.$refs);
    const chatRef = wrapper.vm.$refs.chatRef as HTMLDivElement;
    chatRef.scrollTo = scrollToMock;

    // ACT:
    await wrapper.setProps({
      messages: [...messages, { id: 5, message: 'Hello, World!', isMine: true }],
    });

    // Delay to avoid the false positive:
    await new Promise((resolve) => setTimeout(resolve, 1000));

    expect(scrollToMock).toHaveBeenCalled();
    expect(scrollToMock).toHaveBeenCalledTimes(1);
    expect(scrollToMock).toHaveBeenCalledWith({ behavior: 'smooth', top: expect.any(Number) });
  });
});
