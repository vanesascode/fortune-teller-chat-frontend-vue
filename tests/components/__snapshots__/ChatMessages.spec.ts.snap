// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<ChatMessages tests/> > ChatMessages renders correctly 1`] = `
"<div class="flex-1 overflow-y-auto p-4">
  <div class="flex flex-col space-y-2" id="messagesUp">
    <div class="flex justify-end" id="1">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello, World!</div>
    </div>
    <div class="flex" id="2">
      <div class="bg-gray-300 text-black p-2 rounded-lg max-w-xs"><span class="capitalize">Bye, World!</span>
        <!--v-if-->
      </div>
    </div>
    <div class="flex justify-end" id="3">
      <div class="bg-blue-200 text-black p-2 rounded-lg max-w-xs">Hello, World!</div>
    </div>
    <div class="flex" id="4">
      <div class="bg-gray-300 text-black p-2 rounded-lg max-w-xs"><span class="capitalize">Bye, World!</span>
        <!--v-if-->
      </div>
    </div>
    <!--  v-bind="message" is the short way of this --     :message="message.message"
      :isMine="message.isMine"
      :image="message.image" -->
  </div>
</div>"
`;
